### PR TITLE
cli: git push: exclude new bookmarks tracking other remotes if no --remote set

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Review by martinvonz — APPROVED** [review]
Sounds reasonable. I've never run into this problem, but I rarely use more than one remote. Maybe we can ask on Discord if others have any opinion.

**Comment by jennings** [comment]
Am I understanding this feature correctly?

```shell
$ jj git remote add origin ...
$ jj git remote add upstream ...
$ jj bookmark create foo

# pushes foo to upstream and tracks it
$ jj git push -r @ --remote upstream

# this is the new behavior:
# this does nothing since foo@upstream is tracking
$ jj git push -r @

# pushes foo to origin if the remote is mentioned explicitly
$ jj git push -r @ --remote origin
```

If I'm understanding this right, the goal seems reasonable but using `--remote` as the override seems a little obscure to me. My first guess would be that specifying the bookmark with `-b` was the way to do that.

**Comment by yuja** [comment]
> but using `--remote` as the override seems a little obscure to me.

Yes. That's why I added a hint saying "... Use --remote=origin to push to new remote."

> My first guess would be that specifying the bookmark with `-b` was the way to do that.

That seemed scary. User might expect the bookmark would be pushed to the associated remote as `git push` would do.

File: cli/src/commands/git/fetch.rs [hunk 1]
```diff
@@ -46,6 +46,10 @@ pub struct GitFetchArgs {
     branch: Vec<StringPattern>,
     /// The remote to fetch from (only named remotes are supported, can be
     /// repeated)
+    ///
+    /// This defaults to the `git.fetch` setting. If the default remote is not
```
[the remark this answers is not in the document]
Comment (yuja) [reply cli/src/commands/git/fetch.rs]: Updated the doc to say `that`.

**Comment by SoraTenshi** [comment]
> > but using `--remote` as the override seems a little obscure to me.
> 
> Yes. That's why I added a hint saying "... Use --remote=origin to push to new remote."
> 
> > My first guess would be that specifying the bookmark with `-b` was the way to do that.
> 
> That seemed scary. User might expect the bookmark would be pushed to the associated remote as `git push` would do.

As someone who just recently got around starting to use `jj` i can confirm it.
Bookmarks are in my brain associated with a branch and remote, so using a `jj git push -b main` looks a bit scary, and i even used it first only on a repo i don't really care about where i am the only user of.
So yeah, i can second the scary part of it :)

**Comment by martinvonz** [comment]
> > but using `--remote` as the override seems a little obscure to me.
> 
> Yes. That's why I added a hint saying "... Use --remote=origin to push to new remote."

I suppose we could have a separate flag instead, if people like that better. `--allow-new-tracking` or something.

**Comment by ilyagr** [comment]
(No action needed)

The difference between `jj git push` in repo with one remote, `jj git push` in repo with multiple remotes, and `jj git push --tracked` is becoming a bit confusing.

I don't think it has to be fixed in this PR. Fixing this problem might be worth the added confusion until we figure out a better approach (and I don't have one clearly in mind).

> I suppose we could have a separate flag instead, if people like that better. --allow-new-tracking or something.

I like this line of thinking, but I'm not sure what the flag should be exactly or how to name it. Should it allow new tracking for branches that are not already tracking something? Should it exist in addition to `--tracked` or should it replace it?

**Comment by yuja** [comment]
We can of course add `--allow-new-tracking` (i.e. push local-only or tracking bookmarks by default) or `--allow-new-bookmark` (i.e. push tracking and `--change` bookmarks by default). Implementation-wise, the latter is simpler because we can remove the special case of `@git` tracking bookmarks, but I'm not sure if we want this level of strictness.

My feeling is that `--remote=NAME` is kinda okay as a flag to express user's intent to push new bookmarks to the specified remote.

**Comment by martinvonz** [comment]
> My feeling is that --remote=NAME is kinda okay as a flag to express user's intent to push new bookmarks to the specified remote.

I would expect that `--remote=NAME` is just about restricting the set of remotes to the specified one. I think I would find it surprising if it changed the behavior in other ways.

**Comment by yuja** [comment]
Are we okay to change `jj git push` to not push new bookmarks by default?

* `jj git push -b tracking` updates `tracking@origin`
* `jj git push -b local-only` fails
* `jj git push --new-bookmark -b local-only` creates `local-only@origin` and tracks it
* `jj git push -c @-` creates new `push-xxx@origin` and tracks it (`--new-bookmark` is implied for `push-xxx`)
* `jj git push --all` pushes tracking bookmarks, warns skipped bookmarks (basically the same as `--tracked`)
* `jj git push --all --new-bookmark` pushes all bookmarks

This behavior is easier to explain and implement than `--allow-new-tracking`.

**Comment by ilyagr** [comment]
I like this idea in general, though I'd implement it after the release to give us a month to test it out and to see whether it confuses people. Either in this way or some other, I'd like more people to spend some time trying it out and/or thinking about how well it works for them.

One possibility is to have a `--set` option that acts as your `--new-bookmarks` and `--create` option that only creates. The default behavior could be called `--move`. Then it's analogous to `jj branch` commands.

**Comment by ilyagr** [comment]
> jj git push --all pushes tracking bookmarks, warns skipped bookmarks (basically the same as --tracked)

It might then make sense to rename `--all` to `--tracked`. I'm not sure whether it's better to have `--all` and `--tracked` as we have now or `--tracked` and `--tracked --new-bookmarks` (with an implicit "or" in between).

This PR already has this problem. I'm not sure how bad it is (there is a warning) or whether it should be blocking (especially since I don't have a good fix in mind, and pushing too much is also a problem), but I am worried about people losing access to their work because they do `jj git push --all` on one computer and then, say, drive to work without noticing that a branch didn't get pushed. Perhaps we shouldn't call it `--all` in this case? (For `-b`, this PR avoids any problem by refusing to push anything and printing a red error)

At the moment, I'm not sure.

**Comment by yuja** [comment]
> though I'd implement it after the release to give us a month to test it out and to see whether it confuses people.

Agreed.

> **Update:** It's less clear how branch deletion fits into my concept,

Yep. I wouldn't want to think about the movement on `jj git push`. It's the command to synchronize local changes with the remote.

> This PR already has this problem: `--all` is no longer an all-or-nothing operation (which I think it is before this PR? Let me know if I'm confused about that). I am worried about people losing access to their work because they do `jj git push --all` on one computer and then, say, drive to work without noticing that a branch didn't get pushed. Perhaps we shouldn't call it `--all` in this case? (For `-b`, this PR avoids any problem by refusing to push anything and printing a red error)

I assume the user would notice warnings about skipped bookmarks. If they don't, they wouldn't notice push error either.

I have no idea whether `--all` should imply `--new-bookmark`. In some sense, `--all` can be considered a flag to disable safety checks and push everything.

**Comment by ilyagr** [comment]
> I assume the user would notice warnings about skipped bookmarks. If they don't, they wouldn't notice push error either.

My reasoning is that people pay attention when nothing is pushed and all they see is an error.  If people see the familiar progress bar of branches being pushed, they might not notice the warning or that the branch they cared about was not pushed.

So, I like `--all` being an all-or-nothing operation. All-or-nothing operations are easy to reason about. Partial success requires attention to understand. Sacrificing this property to fix another problem would be just that, a sacrifice. 

Aside: I've considered from time to time also having a `--all --push-what-you-can` option, but people will hopefully pay attention when they use it.

Aside: I'm not sure `--all` is that great of a UI in the first place for pushing anywhere but a personal fork. But I'm not sure how to change it.

**Comment by yuja** [comment]
> Aside: I'm not sure `--all` is that great of a UI in the first place for pushing anywhere but a personal fork. But I'm not sure how to change it.

Maybe deprecate `--all` and suggest `-b glob:*` instead? It should match your intuition. `--all` would skip conflict bookmarks (so user might lose data), but `-b glob:*` wouldn't.

**Comment by ilyagr** [comment]
That's an interesting thought. I don't understand your last point though: what would `-b glob:*` do for conflicted bookmarks?

**Comment by yuja** [comment]
> what would `-b glob:*` do for conflicted bookmarks?

`push` would fail at all (because the specified bookmarks couldn't be pushed.) OTOH, `--all` would emit warning and continue.

**Comment by yuja** [comment]
I'll merge the first two commits (doc + refactor). I'll split the last patch to new PR because I'm going to add an explicit flag to push new bookmarks.

**Pull request merged**